### PR TITLE
Geometry shader support in Vanilla

### PR DESCRIPTION
**Pull request by IMS212**
commit names got messed up, sorry (would probably be smart to squash)

Patching is extremely broken, but the shader gets attached.

### CODE REVIEW

File: src/main/java/net/coderbot/iris/mixin/MixinProgramManager.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import net.coderbot.iris.pipeline.newshader.ExtendedShader;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(ProgramManager.class)
+public class MixinProgramManager {
+	@Inject(method = "releaseProgram", at = @At(value = "INVOKE", target = "Lcom/mojang/blaze3d/platform/GlStateManager;glDeleteProgram(I)V"))
+	private static void closeGeometryShader(Shader shader, CallbackInfo ci) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/MixinProgramManager.java:14]: iris$

File: src/main/java/net/coderbot/iris/mixin/MixinShaderInstance.java [hunk 2]
```diff
@@ -95,4 +96,11 @@
 
 		logger.warn(message, arg1, arg2);
 	}
+
+	@Inject(method = "attachToProgram", at = @At("HEAD"))
+	private void attachGeometryShader(CallbackInfo ci) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/mixin/MixinShaderInstance.java:101]: iris$

File: src/main/java/net/coderbot/iris/pipeline/newshader/ExtendedShader.java [hunk 3]
```diff
@@ -146,4 +163,8 @@ public boolean addDynamicSampler(IntSupplier sampler, String... names) {
 
 		return used;
 	}
+
+	public Program getGeometry() {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/pipeline/newshader/ExtendedShader.java:167]: Probably want either a Nullable annotation or Optional here.

**Review by coderbot16 — CHANGES_REQUESTED** [review]
A few small comments. Also, what do you mean by "patching is extremely broken"?

File: src/main/java/net/coderbot/iris/pipeline/newshader/TriforcePatcher.java [hunk 4]
```diff
@@ -11,6 +11,10 @@
 	public static String patch(String source, ShaderType type, AlphaTest alpha, boolean hasChunkOffset, ShaderAttributeInputs inputs) {
 		// TODO: Only do the NewLines patches if the source code isn't from gbuffers_lines
 
+		if (source == null) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/pipeline/newshader/TriforcePatcher.java:14]: We shouldn't be passing null in the first place

**Comment by IMS212** [comment]
The patcher somehow just gives up and doesn't patch properly, leading to shader compile failures.

**Comment by IMS212** [comment]
Not specifically, but I do know it effects seus ptgi, which patches perfectly with Sodium

**Comment by coderbot16** [comment]
Marking as a draft due to the aforementioned issues & merge conflicts.

**Pull request closed without merging**